### ISSUE TITLE
Consider implementing full copy-on-write

### ISSUE THREAD

**Issue by Nikratio** [s3ql/s3ql#286]
Currently, s3qlcp duplicates each entry in the metadata database so that only storage objects are shared.

The reason for this is that we can't implement copy on write for table rows while also supporting hardlinks: suppose we mark an entry in the inode tabe as copy-on-write and receive a modification request. We create a new row, and now need to update the corresponding entries in the `contents` table. But which ones? We could keep track of how we arrived at the given row (by dynamically creating inodes in lookup(), so that they reflect the path taken), but there is no way to figure out which other (hard-linked) contents entries also need to be updated and which do not.

For the user, the problem presents itself as one of semantics. Suppose we do `s3qlcp orig copy`, and `bla` is hard-linked to `orig/foo`. What should `copy/bla` be hard-linked to?

Perhaps we could define that hard-links from outside the copy source get broken on copy. However, this may still be difficult to implement (we'd have to do some sort of reverse search on every RCW).

But.. what if we dropped support for hard-links? Are these actually used? The upside is that it would be s3qlcp extremely fast and would reduce the size of the metadata file a lot.

**Comment by d--j** [comment]
> [...] But.. what if we dropped support for hard-links? Are these actually used? [...]

I personally do not use hard links but some people might use [`rsnapshot`](https://rsnapshot.org/) or `rysnc` with the [`--link-dest` option](https://download.samba.org/pub/rsync/rsync.1#opt--link-dest) instead of `contrib/s3ql_backup.sh` to do their backups with S3QL. Their backup would break after dropping hard link support. Maybe ask on the mailing list whether hard links are used by anybody?

We at least would need an upgrade path that would convert existing hard links to normal copies.

**Comment by Nikratio** [comment]
Is there any upside of using `rsnapshot` or `rsync --link dest` over `s3qlcp` with regular rsync? Since these use-cases actually break hardlinks in the source tree, to me they look like perfect examples of workflows that would benefit from a no-hardlink S3QL.

**Comment by Nikratio** [comment]
> For the user, the problem presents itself as one of semantics. Suppose we do `s3qlcp orig copy`, and `bla` is hard-linked to `orig/foo`. What should `copy/bla` be hard-linked to?

For comparison, here's how this is currently handled:

- btrfs doesn't support hardlinks across subvolumes, and snapshots can only be taken of subvolumes
- ZFS doesn't support hardlinks across datasets, and snapshots can only be taken of datasets
- `s3qlcp orig copy` will preserve hardlinks within `orig` (since it discovers them during the copy). Hardlinks from files outside of `orig` to within `orig` will remain as-is, with no connection to `copy` (i.e., the st_nlink values in `copy` can be lower than in `orig`).
-

**Comment by d--j** [comment]
> Is there any upside of using `rsnapshot` or `rsync --link dest` over `s3qlcp` with regular rsync? Since these use-cases actually break hardlinks in the source tree, to me they look like perfect examples of workflows that would benefit from a no-hardlink S3QL.

Yes with S3QL the `s3qlcp`/`s3qllock` + normal `rsync` is a better solution than the `rsnapshot` or `rsync --link-dest` variants. They use hard-links for saving HDD space and are not aware of the built-in block-level deduplication in S3QL.

If you don't bother much how to do stuff optimally in S3QL and used e.g. `rsnapshot` and it's powerful tiered expiration strategy before, you might just have plucked these two together and it worked. If we remove hard link support this will stop working. I just wanted to mention this (quite obvious) point. That might just be fine (since nobody might be using this anyway or if they do we advice them to switch their setup to `contrib/s3ql_backup.sh`).


---

BTW: can we support the "standard way" of doing CoW via [`cp --reflink=always`](https://www.man7.org/linux/man-pages/man1/cp.1.html) or [ioctll FICLONE](https://www.man7.org/linux/man-pages/man2/ioctl_ficlonerange.2.html)? As far as I can see pyfuse3 does not have the needed support for ioctl, right?

**Comment by d--j** [comment]
> Actually, I don't think we would need to drop support for hardlinks completely. Allowing them only within the same directory would be sufficient.

But what about moving the hard link after creating it?
```shell
mkdir folder1 folder2
touch folder1/f1
ln folder1/f1 folder1/f2 # f1 and f2 are in the same directory...
mv folder1/f2 folder2/ # ...and now they are not
```

Optionally disabling hard links file system wide as a performance tweak sounds like a good solution to me. This could be a mount option (no need for new file system revision) and we will check if there are any inodes having `st_nlink > 1` before actually mounting the file system.

**Comment by Nikratio** [comment]
> > Actually, I don't think we would need to drop support for hardlinks completely. Allowing them only within the same directory would be sufficient.
> 
> But what about moving the hard link after creating it?
> 
> ```shell
> mkdir folder1 folder2
> touch folder1/f1
> ln folder1/f1 folder1/f2 # f1 and f2 are in the same directory...
> mv folder1/f2 folder2/ # ...and now they are not

We would make the rename() syscall command fail with an error when renaming a hardlinked file into a different directory (similar to how it fails when renaming files across filesystem boundaries). Depending on the error code that we choose, the `mv` command would fall back to copy+delete, or just bail out with an error.

**Comment by Nikratio** [comment]
> BTW: can we support the "standard way" of doing CoW via [`cp --reflink=always`](https://www.man7.org/linux/man-pages/man1/cp.1.html) or [ioctll FICLONE](https://www.man7.org/linux/man-pages/man2/ioctl_ficlonerange.2.html)? As far as I can see pyfuse3 does not have the needed support for ioctl, right?

I am not sure what `cp --reflink` does internally. If it uses an ioctl, then this is may or may not be supported - depending on `FICLONE` being a "restricted" or "unrestricted" ioctl (see https://github.com/libfuse/libfuse/blob/master/include/fuse_lowlevel.h#L1036). On top of my head, I'm not sure what that difference is and how to find out, but if it's unrestricted then extending pyfuse3 and S3QL to support it should be trivial.
